spec/registry.md

```diff
@@ -1,11 +1,34 @@
 # MessageFormat 2.0 Default Function Registry
 
-This section describes the functions for which each implementation MUST provide
-a _function handler_ to be conformant with this specification.
-
-Implementations MAY implement additional _functions_ or additional _options_.
-In particular, implementations are encouraged to provide feedback on proposed
-_options_ and their values.
+This section defines the **standard** _functions_ which are REQUIRED for conformance with this specification,
+along with **optional** _functions_ that SHOULD be implemented to support
+additional functionality.
+
+Implementations MUST define and support each **standard** _function_ and
+MUST accept all _options_ and _option_ values defined as **standard** listed in this specification.
+
+When defined and supported by an implementation, each **optional** _function_ 
+MUST accept all _options_ and _option_ values listed as **standard** in this specification.
+
+Implementations SHOULD implement _options_ and _option_ values that are marked as **optional**.
+
+Implementations MAY implement _functions_ not defined in this specification.
+Such functions MUST use an implementation-defined _namespace_.
```

Comment:
A "MUST" here would make it impossible for an implementation to update a part of the core function set when it changes.

For example, a later LDML release could add some minor option to an existing formatter, and at the same time introduce a new `:list` formatter. If the implementation cannot immediately provide a `:list` formatter, it could not add the minor option either if there's no flexibility on non-namespaced function and option names.

Also, we can't really test for this.

```suggestion
Such functions SHOULD use an implementation-defined _namespace_.
```

Reply:
I'm having trouble understanding your reasoning. It seems to be something like:

- My implementation adds :foo:list with option backwards=
- LDML 50 introduces a :list formatter with different options.
- My implementation can't conformantly upgrade to LDML 50 by aliasing :list to my :foo:list order to add the :list formatter, because backwards= is not standard and doesn't have a namespace.

Is that the scenario you have in mind? I just can't make out what it is.

Reply:
Ok, I understand what you mean now.

NOTE I think in the future we probably want to introduce any new functions/options/option values as _optional_ in their first release, and only make them _required_ in a subsequent release. That does not solve the problem you raise, but reduces the frequency of occurrence.

The downside of your approach is that it opens the door to collisions, and functions that are 'pretending' to be standard but aren't. I would rather see an **explicit** carve-out for that case.

```suggestion
Such functions MUST use an implementation-defined _namespace_. 

There is one exception to this policy, where the function is conformant to a later version of LDML. For example, suppose LDML 50 adds a `:list` function; in that case, an implementation that is conformant to LDML 49 can add that `:list` function and its options, while remaining conformant to LDML 49. 
```
I'll make the similar suggestion below, for options.

Reply:
I get it, but recall that we're going to break the functions apart into separate specs (according to our design doc). I can get behind SHOULD here to avoid versioning issues, but perhaps we could still say something like:

```suggestion
Functions not defined by any version of this specification MUST use an implementation-defined _namespace_.
```

This is still untestable, of course.

Reply:
Much pithier than my suggestion!

>This is still untestable, of course.

Right, it isn't _mechanically_ testable, but it is testable (in a broader sense) by inspection of implementation's documentation of what it accepts for functions, options, and option values.

Reply:
> Much pithier than my suggestion!

Earned my paycheck today. 😉

Reply:
I think I'll commit this one. I'm curious to hear @eemeli's opinion on options, tho... This formulation *could* apply there too.